sycl/doc/extensions/proposed/sycl_ext_oneapi_p2p.asciidoc

```diff
@@ -0,0 +1,157 @@
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_P2P` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Peer to Peer (P2P) Memory Access APIs
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds four new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class p2p_property {
+  access_supported,
+  atomics_supported
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+  int ext_oneapi_get_p2p_property(const device &peer, ext::oneapi::p2p_property value);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+| Member Function | Description
+
+| bool can_access_peer(device peer)
+| Queries if this device may directly access the peer device's memory.
+Returns true if direct access is supported.
+
+| void enable_peer_access(device peer)
+| Enables this device to access USM device allocations located on the peer
+device. This does not permit the peer device to access this device's memory.
+Throws an exception if access cannot be enabled or if access is already
+enabled. This device and the peer device must be in the same context.
+
+| void disable_peer_access(device peer)
+| Disables access to the peer device's memory from this device. Throws an
+exception if access cannot be disabled or if access is not enabled.
+
+| int get_p2p_property(const device &peer, ext::oneapi::p2p_property value)
+| This method allows querying extra details about peer-to-peer connections
+between this device and the peer device.
+Valid properties are defined below. Backends may define additional
+backend-specific properties that may also be queried.    
+
+|===
+
+=== P2P Properties Query
+
+|===
+| ext::oneapi::p2p_property::access_supported
+| Returns 1 if peer access is supported between this device and the peer device.
+Note that querying this property is similar to calling can_access_peer.
+
+| ext::oneapi::p2p_property::atomics_supported
```

Comment:
I think we should find some other word than "property" here to avoid confusion with the compile-time property feature.  I also think we could shorten the names without losing any meaning.  How about:

```
enum class peer_access {
  supported,
  atomic
};
```

Usage would then read like:

```
// Test if peer device supports access at all:
dev.ext_oneapi_can_access_peer(other, ext::oneapi::peer_access::supported);

// ... which is equivalent to:
dev.ext_oneapi_can_access_peer(other);

// Test if peer device supports atomic access:
dev.ext_oneapi_can_access_peer(other, ext::oneapi::peer_access::atomic);

Reply:
My issue here is that not all queries may have boolean results.  Other APIs support a "give me a number that describes how GOOD the link is".

Reply:
OK, how about:

```
enum class peer_access {
  basic,
  atomic
};
```

If we add an API in the future that returns some non-boolean information about the link, the code will read like this:

```
unsigned val = dev.ext_oneapi_get_peer_bandwidth(other, ext::oneapi::peer_access::basic);
```

Reply:
Eh maybe it's better to just make *that* sorta thing a different API.